File: email_stepup.py
from flask import Flask, jsonify
import pandas as pd
import smtplib
import base64
import requests
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import os
app = Flask(__name__)


# Gmail SMTP Configuration
SMTP_SERVER = "smtp.gmail.com"
SMTP_PORT = 587
SENDER_EMAIL = os.getenv('SENDER_EMAIL')  # Replace with your Gmail

def get_access_token():
    """Fetch a new OAuth access token using the refresh token"""
    url = "https://oauth2.googleapis.com/token"
    data = {
        "client_id": os.getenv("CLIENT_ID"),
        "client_secret": os.getenv("CLIENT_SECRET"),
        "refresh_token": os.getenv("REFRESH_TOKEN"),
        "grant_type": "refresh_token"
    }

    response = requests.post(url, data=data)
    if response.status_code == 200:
        return response.json()["access_token"]
    else:
        raise Exception(f"Failed to get access token: {response.text}")

def load_email_template(username, password):
    """Load HTML email template and insert user credentials."""
    with open("./templates/email_template.html", "r", encoding="utf-8") as file:
        template = file.read()
    
    return (template.replace("{username}", str(username)).replace("{password}",str(password) ))

def send_email(to_email, username, password):
    """Send an email using Gmail SMTP with OAuth 2.0"""
    try:
        access_token = get_access_token()

        # Email subject
        subject = "Welcome to Udaan UPSC – Your Login Credentials"

        # HTML Email Body
        body =  load_email_template(username, password)
        # Create email message
        msg = MIMEMultipart()
        msg["From"] = SENDER_EMAIL
        msg["To"] = to_email
        msg["Subject"] = subject
        msg.attach(MIMEText(body, "html"))  # Send email as HTML

        # Connect to Gmail SMTP server using OAuth 2.0
        server = smtplib.SMTP(SMTP_SERVER, SMTP_PORT)
        server.starttls()
        auth_string = f"user={SENDER_EMAIL}\x01auth=Bearer {access_token}\x01\x01"
        server.docmd("AUTH", f"XOAUTH2 {base64.b64encode(auth_string.encode()).decode()}")
        server.sendmail(SENDER_EMAIL, to_email, msg.as_string())
        server.quit()

        return f"Email sent to {to_email}"

    except Exception as e:
        return f"Failed to send email to {to_email}. Error: {str(e)}"

File: email_stepup.php
<?php
require 'vendor/autoload.php'; // Composer autoload

use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\Exception;
use PHPMailer\PHPMailer\OAuth;
use League\OAuth2\Client\Provider\Google; // Import the Google provider

const SMTP_SERVER = "smtp.gmail.com";
const SMTP_PORT = 587;
$sender_email = "rahul.11919011621@ipu.ac.in";

function get_access_token() {
    $url = "https://oauth2.googleapis.com/token";
    $data = [
        "client_id" => "demo",
        "client_secret" => "demo",
        "refresh_token" => "demo",
        "grant_type" => "refresh_token"
    ];

    $options = [
        'http' => [
            'header' => "Content-Type: application/x-www-form-urlencoded\r\n",
            'method' => 'POST',
            'content' => http_build_query($data),
        ],
    ];
    
    $context = stream_context_create($options);
    $response = file_get_contents($url, false, $context);
    
    if ($response === false) {
        throw new Exception("Failed to get access token");
    }
    
    $json_response = json_decode($response, true);
    if (isset($json_response['access_token'])) {
        return $json_response['access_token'];
    } else {
        throw new Exception("Failed to get access token: " . $response);
    }
}

function load_email_template($username, $password) {
    $template = file_get_contents("./templates/email_template.html");
    if ($template === false) {
        throw new Exception("Failed to load email template");
    }
    
    $template = str_replace("{username}", strval($username), $template);
    $template = str_replace("{password}", strval($password), $template);
    
    return $template;
}

function send_email($to_email, $username, $password) {
    global $sender_email;
    
    try {
        $access_token = get_access_token();
        $subject = "Welcome to Udaan UPSC – Your Login Credentials";
        $body = load_email_template($username, $password);

        $mail = new PHPMailer(true);
        $mail->isSMTP();
        $mail->Host = SMTP_SERVER;
        $mail->SMTPAuth = true;
        $mail->SMTPSecure = 'tls';
        $mail->Port = SMTP_PORT;
        $mail->AuthType = 'XOAUTH2';
        
        $mail->setOAuth(
            new OAuth([
                'provider' => new Google([
                    'clientId' => "696722818906-si3am383fsavh4irf5im892aak93tenc.apps.googleusercontent.com",
                    'clientSecret' => "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",
                ]),
                'clientId' => "696722818906-si3am383fsavh4irf5im892aak93tenc.apps.googleusercontent.com",
                'clientSecret' => "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",
                'refreshToken' => "1//04AMLLFsXNPkmCgYIARAAGAQSNwF-L9IrUiy9RmtHLINDIY5hXaxZQmlOGS5-PCEKphCq0ST2n9Jb1z7L2jPv8L2oXQItGgC6Al4",
                'userName' => $sender_email,
                'accessToken' => $access_token,
            ])
        );

        $mail->setFrom($sender_email);
        $mail->addAddress($to_email);
        $mail->Subject = $subject;
        $mail->isHTML(true);
        $mail->Body = $body;

        $mail->send();
        return "Email sent to {$to_email}";
    } catch (Exception $e) {
        return "Failed to send email to {$to_email}. Error: " . $e->getMessage();
    }
}
?>
